src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioDiagnosticAnalyzerService.cs

```diff
@@ -74,5 +128,155 @@ public VisualStudioDiagnosticAnalyzerService(VisualStudioWorkspace workspace, ID
+        private void OnRunCodeAnalysisForSelectedProjectStatus(object sender, EventArgs e)
+        {
+            var command = (OleMenuCommand)sender;
+
+            // We hook up the "Run Code Analysis" menu commands for CPS based managed projects.
+            // These commands are already hooked up for csproj based projects in StanCore, but those will eventually go away.
+            var visible = TryGetSelectedProjectHierarchy(out var hierarchy) &&
+                hierarchy.IsCapabilityMatch("CPS") &&
+                hierarchy.IsCapabilityMatch(".NET");
+            var enabled = false;
+
+            if (visible)
+            {
+                if (command.CommandID.ID == RunCodeAnalysisForSelectedProjectCommandId &&
+                    hierarchy.TryGetProject(out var project))
+                {
+                    // Change to show the name of the project as part of the menu item display text.
+                    command.Text = string.Format(ServicesVSResources.Run_Code_Analysis_on_0, project.Name);
+                }
+
+                enabled = !IsBuildActive();
+            }
+
+            if (command.Visible != visible)
+            {
+                command.Visible = visible;
+            }
+            if (command.Enabled != enabled)
+            {
+                command.Enabled = enabled;
+            }
+        }
+
+        private void OnRunCodeAnalysisForSelectedProject(object sender, EventArgs args)
+        {
+            if (TryGetSelectedProjectHierarchy(out var hierarchy))
+            {
+                RunAnalyzers(hierarchy);
+            }
+        }
+
+        public void RunAnalyzers(IVsHierarchy? hierarchy)
+        {
+            var project = GetProject(hierarchy);
+            var projectOrSolutionName = project?.Name ?? PathUtilities.GetFileName(_workspace.CurrentSolution.FilePath) ?? "Solution";
+
+            // Add a message to VS status bar that we are running code analysis.
+            var waitDialogMessage = string.Format(ServicesVSResources.Running_code_analysis_for_0, projectOrSolutionName);
+            var statusBar = _serviceProvider?.GetService(typeof(SVsStatusbar)) as IVsStatusbar;
+            statusBar?.SetText(waitDialogMessage);
+
+            // Force complete analyzer execution in background.
+            Task.Run(async () =>
+            {
+                await _diagnosticService.ForceAnalyzeAsync(_workspace.CurrentSolution, project?.Id, CancellationToken.None).ConfigureAwait(false);
+
+                // Add a message to VS status bar that we completed executing code analysis.
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+                var notificationMesage = string.Format(ServicesVSResources.Code_analysis_completed_for_0_Please_wait_for_the_error_list_to_refresh, projectOrSolutionName);
+                statusBar?.SetText(notificationMesage);
+            });
+        }
+
+        private Project? GetProject(IVsHierarchy? hierarchy)
```

Comment:
📝 This is also implemented for Code Cleanup. Consider consolidating the implementations.

Reply:
This functionality seems to be used at bunch of places. I think the best place to move this would be [IVsHierarchyExtensions](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/Extensions/IVsHierarchyExtensions.cs,57083cc56e24e14b). I will do the following as a following PR:
1. Enable nullable on IVsHierarchyExtensions
2. Move this helper in that file
3. Switch all clients to use this helper instead of cloning functionality.